integration-tests/tests/src/tests/sync/mixed.ts

```diff
@@ -15,29 +15,181 @@
+  };
+}
+
+const bool = true;
+const int = 1;
+const double = 123.456;
+const d128 = BSON.Decimal128.fromString("6.022e23");
+const string = "hello";
+const date = new Date();
+const oid = new BSON.ObjectId();
+const uuid = new BSON.UUID();
+const nullValue = null;
+const data = new Uint8Array([0xd8, 0x21, 0xd6, 0xe8, 0x00, 0x57, 0xbc, 0xb2, 0x6a, 0x15]).buffer;
+
+function getMixedList(realm: Realm) {
+  const expectedObj = { _id: new BSON.ObjectId() };
+  const obj = realm.create(MixedClass, expectedObj);
+
+  const values = [bool, int, double, d128, string, oid, uuid, nullValue, date, data, obj];
+  const expected = [bool, int, double, d128, string, oid, uuid, nullValue, date, data, expectedObj];
+
+  return { values, expected };
+}
+
+function getMixedDict(realm: Realm) {
+  const expectedObj = { _id: new BSON.ObjectId() };
+  const obj = realm.create(MixedClass, expectedObj);
+
+  const values = {
+    bool,
+    int,
+    double,
+    d128,
+    string,
+    oid,
+    uuid,
+    nullValue,
+    date,
+    data,
+    obj,
+  };
+
+  const expected = {
+    bool,
+    int,
+    double,
+    d128,
+    string,
+    oid,
+    uuid,
+    nullValue,
+    date,
+    data,
+    obj: expectedObj,
+  };
+
+  return { values, expected };
+}
+
+function getNestedMixedList(realm: Realm) {
+  const mixList1 = getMixedList(realm);
+  const mixList2 = getMixedList(realm);
+  const mixDict = getMixedDict(realm);
+
+  const values = [...mixList1.values, mixList2.values, mixDict.values];
+  const expected = [...mixList1.expected, mixList2.expected, mixDict.expected];
+
+  return { values, expected };
+}
+
+function getNestedMixedDict(realm: Realm) {
+  const mixDict1 = getMixedDict(realm);
+  const mixDict2 = getMixedDict(realm);
+  const mixList = getMixedList(realm);
+
+  const values = {
+    ...mixDict1.values,
+    innerDict: mixDict2.values,
+    innerList: mixList.values,
+  };
+
+  const expected = {
+    ...mixDict1.expected,
+    innerDict: mixDict2.expected,
+    innerList: mixList.expected,
+  };
+
+  return { values, expected };
+}
+
+function expectRealmList(value: unknown): asserts value is Realm.List<unknown> {
+  expect(value).instanceOf(Realm.List);
+}
+
+function expectRealmDictionary(value: unknown): asserts value is Realm.Dictionary<unknown> {
+  expect(value).instanceOf(Realm.Dictionary);
+}
 
 /**
  * The default tester of values.
  * @param actual The value downloaded from the server.
  * @param inserted The value inserted locally before upload.
  */
-function defaultTester(actual: Realm.Mixed, inserted: Realm.Mixed) {
-  expect(actual).equals(inserted);
+function defaultTester(actual: unknown, inserted: unknown) {
+  if (inserted instanceof Array) {
+    expectRealmList(actual);
+    expect(actual.length).equals(inserted.length);
+    for (let index = 0; index < actual.length; index++) {
+      defaultTester(actual[index], inserted[index]);
+    }
+  } else if (inserted != null && typeof inserted === "object" && "d128" in inserted) {
```

Comment:
We could check for an `Object` constructor to enter this for the objects you create using e.g. `{}`. Was that the intention with `d128`?
```suggestion
  } else if (inserted !== null && typeof inserted === "object" && inserted.constructor === Object) {
```

Nit: If you want to get rid off the type assertion to `Record` further below, an alternative is to pull that condition out to e.g.:
```ts
function isJsObject(value: unknown): value is Record<string, unknown> {
  return value !== null && typeof value === "object" && !Array.isArray(value) && value.constructor === Object;
}
```

Then the else-if could use:
```ts
} else if (isJsObject(inserted)) {
  // ...
  for (const key of insertedKeys) {
    defaultTester(actual[key], inserted[key]);
    // Instead of:
    // defaultTester(actual[key], (inserted as Record<string, unknown>)[key]);
  }
} // ...
```

But since it'd only be used there it's not super necessary imo.

Reply:
Yes, I'm doing it because I want to differentiate between a "dictionary" mixed value and an "object" mixed value, and I didn't think of anything more clever than this 😁 
I'll add a comment.